### ISSUE TITLE
Version 1.6 values vs global 

### ISSUE THREAD

**Issue by tommyJimmy87** [tazjin/kontemplate#142]
I tried to update to 1.6 version and i noticed that when i try to overwrite one variables, that i have as global variable, inside a single component this doesn't make effect and the global variables win against the included variable within the components.

This could be the problem but i'm not sure.
https://github.com/tazjin/kontemplate/commit/77ca5b47cf5d40fa53204633924e8ba30a517b3f

**Comment by tazjin** [comment]
Hey!

From my perspective that commit was actually a bug fix (i.e. the previous behaviour was conceptually wrong). Can you tell me how you use the globals and in what cases you want the resource set defaults to override them?

I think there may be a use-case here I haven't quite understood yet :)

**Comment by tommyJimmy87** [comment]
Here an example : 

```
---
context: some-context 
global:
  namespace: global-namespace
include:
  - name: some-components
    path: ../some-components
    include:
      - name: component1
        path: component1/apply
        values:
           namespace: component1-namespace
       - name: component2
          path: component2/apply
```

Within the component, the namespace definition as variable is :  **{{.namespace}}** , the component's content is not actually important.

In this case with the previous version ( **until 1.5** ) **component1** namespace will be : **component1-namespace** , instead **component2** namespace will be remain : **global-namespace**.

Now for me this is correct, because i have the globally variables for all my components and if i want to replace one of the globally i just have to add that variable as **values** inside my component. 

In the **1.6** version instead also for the **component1** will be remain the global one : **global-namespace**. 

In my case i have a bunch of components defined and not all components have the same global values, and if i'll rollout the new version i have to convert all my global variables in component variables. Anyway I need something replacing the global variables with custom variables. I think this is very useful.

Thanks

**Comment by tazjin** [comment]
@tommyJimmy87 Thanks for the explanation. I think there is some ambiguity with the precedence of defaults vs. globals and maybe it should simply be a toggleable behaviour. I'll think about this a bit more after work :)

**Comment by Arttii** [comment]
I think it make sense to have a way to define global variables to reused across components. Another very simple example would say if you're using a certificate for all elbs you have and they are different per environment with dozens of components. So having 20 times repeated
```
cert: arn:aws:acm:eu-central-1:12322312:certificate/4sdsdadawe213232-223238532c
```
is not super optimal I think.

**Comment by tazjin** [comment]
@Arttii You can already use variables under the `global` key in a context file, those will be available to all resource sets imported below.

The discussion here is mostly about the precedence order in which variables with the same name at different levels override each other.

My mental model (so far) has been that it should be like this:

```
CLI variables > variables in `include` section > imports > globals > defaults
```

However due to a bug in the logic before 1.6 the order was

```
CLI variables > variables in `include` section > imports > defaults > globals
```

But @tommyJimmy87 highlighted a use-case where this is actually the *correct* order, so it's not as unambiguous as I thought initially.

I'm experimenting with a few different things but there'll probably be some kind of toggle for this eventually.

**Comment by Arttii** [comment]
@tazjin I see. But to be honest for me global varaibles are there to override defaults, which are there in case no other thing was specified. But maybe its just me.

**Comment by tazjin** [comment]
@Arttii Aren't we saying the same thing then? :thinking:

**Comment by Arttii** [comment]
Hmm I am also confused now. I will think about it a bit more properly tomorrow when tommy is also in the office.
________________________________

**Comment by merlineus** [comment]
I'm also confused with this behaviour after updating to 1.6. My configuration have a global "replicas" variable to set default replicas count (i.e. 1) for all resources. And some resources have an overrided value (for example 3). The same way I've set tag for images - "latest" by default and exact versions for separate resources. So this logic doesn't work anymore :)

**Comment by tazjin** [comment]
^ Please take a look at the PR above.

**Comment by Arttii** [comment]
So this i very confusing, our problem was that `values` supplied in the component explicitly
```
global:
  namespace: beta
- name: grafana
   path: grafana/apply
   values:
           component: grafana
           namespace: dev
```
were being overshadowed by Global and not the default being overshadowed by global. I think its completely fine for defaults to be overshadowed by Globals. In this example the grafana component does not even have a default for namespace. Are those default values or what? Because I thoughts defaults are the values specified in `default.yaml`.

**Comment by merlineus** [comment]
The order 

`CLI variables > variables in `include` section > imports > globals > defaults`

seem reasonable, if I correctly understand that variables on the right side overriding variables on the left side. Maybe only "imports" and "globals" could be swapped because from my POV (and maybe shadowing logic) included file should have less priority in hierarchy.

But I can't figure out how swapping "globals" and "defaults" in the chain affected "include" section which should override all definitions  to the right of it in chain?

**Comment by Arttii** [comment]
@merlineus Basically the vars in ```values:``` are treated as defaults. So for our usecase we need something like:

```CLI variables > variables in include section > imports >explicit> globals > defaults``` or 
```CLI variables > explicit> variables in include section > imports > globals > defaults```

**Comment by merlineus** [comment]
@Arttii I thought "values:" are "variables in include section"... Now I'm confused even more :)

**Comment by Arttii** [comment]
So my example up top ```CLI variables``` and ```include variables``` would be empty. And ```defaults``` would be ```component: grafana,namespace: dev```, which will be merged with ```namespace: beta```. Maybe easy fix with be to merge stuff in ```values:``` with included vars?

**Comment by merlineus** [comment]
The simplest (and logical for me) order is:
`CLI variables > variables in include section ("values:" in the resource section) > globals`

In that case we have one "global" section which could be overridden by resource "values:" section which could be overridden by CLI definitions.

"Defaults" are only the predefined defaults for resource which could be overridded on any level (useful for distributing resources ready for templating).

And "imports" used only for structuring the resource set allowing to export some part of "global" or "values:" section to external file (but should have less priority than variables in section which contains "import").

**Comment by Arttii** [comment]
Yep, that would work for us as well

**Comment by tazjin** [comment]
If you guys have time, please take a look at the pull request above (specifically the new implementation of value merging in `context.go`) and test contained release candidate (or build a release yourself if you prefer that!) for 1.7.0.

The only logical change is that values from the `import` section have lower precedence than `globals`, based on what @merlineus said.

I think this makes sense because things that are less specific to a particular execution of kontemplate have lower precedence, and import variables are potentially less specific because they may be used in multiple contexts.

To quote my own comment, the order is now (in ascending priority):

1. Default values in resource sets (via `default.yaml` / `default.json`)
2. Values imported from files (via `import:`)
3. Global values in a cluster configuration (via `global:`)
4. Values set in a resource set's `include`-section
5. Explicit values set on the CLI (via `--var`)

**Comment by merlineus** [comment]
Yeah!
Thanks for quick reaction. This change sounds reasonable and clear for understanding :)

**Comment by tazjin** [comment]
@tommyJimmy87 Do you consider this issue solved with the recent changes?

**Comment by tazjin** [comment]
Great, closing issue! I'll do the 1.7.0 release soon (tm)

**Comment by tazjin** [comment]
The fix for this is included in [version 1.7.0](https://github.com/tazjin/kontemplate/releases/tag/v1.7.0).
